src/games/strategy/triplea/ai/proAI/util/ProMoveOptionsUtils.java

```diff
@@ -1163,8 +1175,77 @@ private void findBombardOptions(final PlayerID player, final List<Territory> myU
+        isEnemyCapitalOrFactory = true;
+      }
+      if (patd.getMaxBattleResult().getWinPercentage() < MIN_WIN_PERCENTAGE && isEnemyCapitalOrFactory
+          && alliedAttackOptions.getMax(t) != null) {
+
+        // Check for allied attack
+        final ProTerritory alliedAttack = alliedAttackOptions.getMax(t);
+        final Set<Unit> alliedUnits = new HashSet<Unit>(alliedAttack.getMaxUnits());
+        alliedUnits.addAll(alliedAttack.getMaxAmphibUnits());
+        if (!alliedUnits.isEmpty()) {
+
+          // Get max enemy defenders
+          final Set<Unit> additionalEnemyDefenders = new HashSet<Unit>();
+          final PlayerID alliedPlayer = alliedUnits.iterator().next().getOwner();
+          final List<PlayerID> players = utils.getOtherPlayersInTurnOrder(player);
+          for (final ProTerritory enemyDefendOption : enemyDefendOptions.getAll(t)) {
+            final Set<Unit> enemyUnits = new HashSet<Unit>(enemyDefendOption.getMaxUnits());
+            enemyUnits.addAll(enemyDefendOption.getMaxAmphibUnits());
+            if (!enemyUnits.isEmpty()) {
+              final PlayerID enemyPlayer = enemyUnits.iterator().next().getOwner();
+              if (utils.isPlayersTurnFirst(players, enemyPlayer, alliedPlayer)) {
+                additionalEnemyDefenders.addAll(enemyUnits);
+              }
+            }
+          }
+
+          // Check allied result without strafe
+          final Set<Unit> enemyDefendersBeforeStrafe = new HashSet<Unit>(defenders);
+          enemyDefendersBeforeStrafe.addAll(additionalEnemyDefenders);
+          final ProBattleResult result =
+              battleUtils.estimateAttackBattleResults(alliedPlayer, t, new ArrayList<Unit>(alliedUnits),
+                  new ArrayList<Unit>(enemyDefendersBeforeStrafe), alliedAttack.getMaxBombardUnits());
+          if (result.getWinPercentage() < WIN_PERCENTAGE) {
+            patd.setStrafing(true);
+
+            // Try to strafe to allow allies to conquer territory
+            final Set<Unit> combinedUnits = new HashSet<Unit>(patd.getMaxUnits());
+            combinedUnits.addAll(patd.getMaxAmphibUnits());
+            final ProBattleResult strafeResult =
+                battleUtils.callBattleCalculator(player, t, new ArrayList<Unit>(combinedUnits), defenders,
+                    patd.getMaxBombardUnits(), true, true);
+
+            // Check allied result with strafe
+            final Set<Unit> enemyDefendersAfterStrafe = new HashSet<Unit>(strafeResult.getAverageDefendersRemaining());
+            enemyDefendersAfterStrafe.addAll(additionalEnemyDefenders);
+            patd.setMaxBattleResult(battleUtils.estimateAttackBattleResults(alliedPlayer, t, new ArrayList<Unit>(
+                alliedUnits), new ArrayList<Unit>(enemyDefendersAfterStrafe), alliedAttack.getMaxBombardUnits()));
+
+
+            ProLogger.debug("Checking strafing territory: " + t + ", alliedPlayer="
```

Comment:
As a comment, it looks like effectively this debug statement is printing out information so you can serve as a human Junit test. The lack of tests, which allows for such long methods I think is the main cause and reason for debug like this to be needed. Said in another way, if this bit of functionality were well encapsulated and tested, it would be rock solid and this debug would not be very interesting.

A more advanced debug would aggregate a stats object per each territory and then give a single consolidated report about decisions. So I don't think it's quite either the case that this is useful application level logging, it is arguably a half step below and is debug / manual testing.